File: mtui/commands/localrun.py
"""The `lrun` command."""

import subprocess
from argparse import REMAINDER
from logging import getLogger

from ..cli.argparse import ArgumentParser
from . import Command

logger = getLogger("mtui.commands.lrun")


class LocalRun(Command):
    """Runs a command in the local shell.

    The command is run in the current working directory where mtui was
    started, unless chroot to the template directory is enabled.

    When the surrounding prompt is interactive (a human at the REPL), the
    child process inherits the terminal so output streams live. Under a
    non-interactive prompt (the MCP server, headless callers), stdout and
    stderr are captured and re-emitted through ``self.sys`` so the caller
    can see them; on non-zero exit the real return code is propagated via
    ``self.sys.exit`` instead of being collapsed to a generic failure.
    """

    command = "lrun"

    @classmethod
    def _add_arguments(cls, parser: ArgumentParser) -> None:
        """Adds arguments to the command's argument parser."""
        parser.add_argument(
            "command", nargs=REMAINDER, help="command to run on local shell"
        )

    def __call__(self) -> None:
        """Executes the `lrun` command."""
        if not self.args.command:
            logger.error("Missing argument")
            return

        cmd = " ".join(self.args.command)

        # Default to the safe (current) streaming behaviour if a caller
        # forgets to set ``interactive`` on the prompt.
        if getattr(self.prompt, "interactive", True):
            subprocess.check_call(cmd, shell=True)
            return

        result = subprocess.run(
            cmd,
            shell=True,
            capture_output=True,
            text=True,
            check=False,
        )
        if result.stdout:
            self.sys.stdout.write(result.stdout)
        if result.stderr:
            self.sys.stderr.write(result.stderr)
        if result.returncode != 0:
            self.sys.exit(result.returncode)

File: crates/mtui-core/src/commands/localrun.rs
//! The `lrun` command.

use async_trait::async_trait;
use clap::{Arg, ArgMatches};

use crate::command::{Command, Scope};
use crate::error::{CommandError, CommandResult};
use crate::session::Session;

/// Runs a command in the local shell.
///
/// Ports upstream `mtui.commands.localrun.LocalRun`. The command runs in mtui's
/// current working directory. When the session is interactive (a human at the
/// REPL) the child inherits the terminal so output streams live; under a
/// non-interactive session (MCP, headless callers) stdout and stderr are
/// captured and re-emitted through the display, and a non-zero exit is surfaced
/// as a command error carrying the real return code.
///
/// The positional tokens are re-quoted with `shlex::join` so a token containing
/// shell metacharacters keeps its quoting instead of being re-split by the
/// `sh -c` shell; shell operators (pipes, redirection) belong inside an explicit
/// `sh -c '...'`.
pub struct LocalRun;

#[async_trait]
impl Command for LocalRun {
    fn name(&self) -> &'static str {
        "lrun"
    }

    fn scope(&self) -> Scope {
        // Local execution never touches hosts or templates: run exactly once.
        Scope::Single
    }

    fn configure(&self, cmd: clap::Command) -> clap::Command {
        cmd.arg(
            Arg::new("command")
                .num_args(0..)
                .trailing_var_arg(true)
                .allow_hyphen_values(true)
                .value_name("COMMAND")
                .help("command to run on local shell"),
        )
    }

    async fn call(&self, session: &mut Session, args: &ArgMatches) -> CommandResult {
        let tokens: Vec<String> = args
            .get_many::<String>("command")
            .map(|it| it.cloned().collect())
            .unwrap_or_default();
        if tokens.is_empty() {
            return Err(CommandError::Other("Missing argument".to_owned()));
        }
        let cmd = shlex::try_join(tokens.iter().map(String::as_str))
            .map_err(|e| CommandError::Other(format!("invalid command: {e}")))?;

        if session.interactive {
            let status = tokio::process::Command::new("sh")
                .arg("-c")
                .arg(&cmd)
                .status()
                .await
                .map_err(|e| CommandError::Other(e.to_string()))?;
            if !status.success() {
                return Err(CommandError::Other(format!(
                    "local command failed: {status}"
                )));
            }
            return Ok(());
        }

        let output = tokio::process::Command::new("sh")
            .arg("-c")
            .arg(&cmd)
            .output()
            .await
            .map_err(|e| CommandError::Other(e.to_string()))?;
        let stdout = String::from_utf8_lossy(&output.stdout);
        let stderr = String::from_utf8_lossy(&output.stderr);
        for line in stdout.lines() {
            session.display.println(line);
        }
        for line in stderr.lines() {
            session.display.println(line);
        }
        if !output.status.success() {
            let code = output.status.code().unwrap_or(-1);
            return Err(CommandError::Other(format!(
                "local command exited with code {code}"
            )));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::commands::testkit::{empty_session, matches};

    #[test]
    fn name_and_single_scope() {
        assert_eq!(LocalRun.name(), "lrun");
        assert_eq!(LocalRun.scope(), Scope::Single);
    }

    #[tokio::test]
    async fn missing_argument_errors() {
        let (mut session, _buf) = empty_session();
        let args = matches(&LocalRun, &[]);
        let err = LocalRun.call(&mut session, &args).await.unwrap_err();
        assert!(matches!(err, CommandError::Other(m) if m == "Missing argument"));
    }

    #[tokio::test]
    async fn captures_stdout_when_headless() {
        // A non-interactive session captures and re-emits child output.
        let (mut session, buf) = empty_session();
        assert!(!session.interactive);
        let args = matches(&LocalRun, &["printf", "hello"]);
        LocalRun.call(&mut session, &args).await.unwrap();
        assert!(buf.contents().contains("hello"), "{}", buf.contents());
    }

    #[tokio::test]
    async fn nonzero_exit_propagates_headless() {
        let (mut session, _buf) = empty_session();
        let args = matches(&LocalRun, &["sh", "-c", "exit 3"]);
        let err = LocalRun.call(&mut session, &args).await.unwrap_err();
        assert!(matches!(err, CommandError::Other(m) if m.contains("code 3")));
    }
}
